helix-term/src/ui/editor.rs

```diff
@@ -735,8 +742,38 @@ impl EditorView {
                 // first execute whatever put us into insert mode
                 self.last_insert.0.execute(cxt);
                 // then replay the inputs
-                for &key in &self.last_insert.1.clone() {
-                    self.insert_mode(cxt, key)
+                for key in self.last_insert.1.clone() {
+                    match key {
+                        InsertEvent::Key(key) => self.insert_mode(cxt, key),
+                        InsertEvent::CompletionApply(compl) => {
+                            let (view, doc) = current!(cxt.editor);
+
+                            doc.restore(view.id);
+
+                            let text = doc.text().slice(..);
+                            let cursor = doc.selection(view.id).primary().cursor(text);
+
+                            let shift_position =
+                                |pos: usize| -> usize { pos + cursor - compl.trigger_offset };
+
+                            let tx = Transaction::change(
+                                doc.text(),
+                                compl
+                                    .transaction
+                                    .changes_iter()
+                                    .filter_map(|(start, end, t)| {
+                                        (start..=end).contains(&compl.trigger_offset).then(|| {
+                                            (shift_position(start), shift_position(end), t)
+                                        })
+                                    }),
```

Comment:
> The only thing that bothers me is doing this filter_map on every replay. Maybe there is some way to cache that.

Let's do the filter_map inside `ui::completion` and store a `Vec<Change>` on the CompleteAction instead.

Reply:
Ok